File: reference/typescript/src/tessellate.ts
import { SSKError } from './error.js';
import type { MeshData, Point, ResolvedPiece, Vec3 } from './types.js';
import {
  add,
  cross,
  cubicBezier,
  cubicBezierDeriv,
  dot,
  interpolateRotation,
  interpolateSize,
  matVec,
  minimalRotation,
  normalize,
  norm,
  projectOntoPlane,
  rotationMatrixXYZ,
  scale,
  solveTransition,
  sub,
  type Mat3,
  type V3,
} from './vecmath.js';

const DEFAULT_RESOLUTION = 32;
const MIN_RESOLUTION = 3;

type Frame = { lx: V3; ly: V3; lz: V3 };
type Ring = { pos: V3; frame: Frame; size: [number, number, number]; rot: [number, number, number]; v: V3[] };
type Segment = {
  n: number;
  pos: (u: number) => V3;
  tan: (u: number) => V3;
  size: (u: number) => [number, number, number];
  rot: (u: number) => [number, number, number];
};

export function tessellate(piece: ResolvedPiece, options: { resolution?: number } = {}): MeshData | null {
  const resolution = checkResolution(options.resolution ?? DEFAULT_RESOLUTION);
  if (piece.points.length === 1) return pointDefined(piece, resolution);
  return pathDefined(piece, resolution);
}

function checkResolution(resolution: number): number {
  if (!Number.isInteger(resolution)) throw new SSKError('resolution must be an integer');
  if (resolution < MIN_RESOLUTION) throw new SSKError(`resolution must be >= ${MIN_RESOLUTION}`);
  return resolution;
}

function effectiveSize(point: Point, piece: ResolvedPiece): [number, number, number] {
  const size = point.size ?? piece.size;
  return [Number(size.x), Number(size.y), Number(size.z)];
}

function effectiveRotation(point: Point, piece: ResolvedPiece): [number, number, number] {
  const rotation = point.rotation ?? piece.rotation ?? { x: 0, y: 0, z: 0 };
  return [Number(rotation.x), Number(rotation.y), Number(rotation.z)];
}

function degenerate(sx: number, sy: number, sz: number): boolean {
  return [sx, sy, sz].filter((value) => value === 0).length >= 2;
}

function crossN(piece: ResolvedPiece, resolution: number): number {
  return piece.shape === 'ngon' ? piece.sides ?? resolution : resolution;
}

function pointDefined(piece: ResolvedPiece, resolution: number): MeshData | null {
  const point = piece.points[0];
  const [sx, sy, sz] = effectiveSize(point, piece);
  if (degenerate(sx, sy, sz)) return null;
  const pos: V3 = [point.x, point.y, point.z];
  const rotation = rotationMatrixXYZ(...effectiveRotation(point, piece));
  if (piece.shape === 'circle') return ellipsoid(pos, sx, sy, sz, rotation, resolution);
  return bipyramid(pos, sx, sy, sz, rotation, piece.sides!);
}

function ellipsoid(c: V3, sx: number, sy: number, sz: number, rotation: Mat3, n: number): MeshData {
  const nLat = Math.max(4, Math.floor(n / 2));
  const nLon = Math.max(6, n);
  const vertices: V3[] = [add(c, matVec(rotation, [0, 0, sz]))];
  for (let i = 1; i < nLat; i += 1) {
    const phi = Math.PI * i / nLat;
    const sp = Math.sin(phi);
    const cp = Math.cos(phi);
    for (let j = 0; j < nLon; j += 1) {
      const th = 2 * Math.PI * j / nLon;
      vertices.push(add(c, matVec(rotation, [sx * sp * Math.cos(th), sy * sp * Math.sin(th), sz * cp])));
    }
  }
  vertices.push(add(c, matVec(rotation, [0, 0, -sz])));

  const faces: number[][] = [];
  for (let j = 0; j < nLon; j += 1) faces.push([0, 1 + j, 1 + ((j + 1) % nLon)]);
  for (let i = 0; i < nLat - 2; i += 1) {
    const a0 = 1 + i * nLon;
    const b0 = 1 + (i + 1) * nLon;
    for (let j = 0; j < nLon; j += 1) {
      const jn = (j + 1) % nLon;
      faces.push([a0 + j, b0 + j, a0 + jn]);
      faces.push([a0 + jn, b0 + j, b0 + jn]);
    }
  }
  const bottom = 1 + (nLat - 1) * nLon;
  const last = 1 + (nLat - 2) * nLon;
  for (let j = 0; j < nLon; j += 1) faces.push([last + j, bottom, last + ((j + 1) % nLon)]);
  return { vertices, faces };
}

function bipyramid(c: V3, sx: number, sy: number, sz: number, rotation: Mat3, sides: number): MeshData {
  const vertices: V3[] = [add(c, matVec(rotation, [0, 0, sz])), add(c, matVec(rotation, [0, 0, -sz]))];
  for (let i = 0; i < sides; i += 1) {
    const angle = 2 * Math.PI * i / sides;
    vertices.push(add(c, matVec(rotation, [sx * Math.cos(angle), sy * Math.sin(angle), 0])));
  }
  const faces: number[][] = [];
  for (let i = 0; i < sides; i += 1) {
    const next = (i + 1) % sides;
    faces.push([0, 2 + i, 2 + next]);
    faces.push([1, 2 + next, 2 + i]);
  }
  return { vertices, faces };
}

function pathDefined(piece: ResolvedPiece, resolution: number): MeshData | null {
  const cn = crossN(piece, resolution);
  const segments: Segment[] = [];
  for (let i = 0; i < piece.points.length - 1; i += 1) {
    const segment = seg(piece.points, i, piece, resolution);
    if (segment) segments.push(segment);
  }
  if (!segments.length) return null;

  const rings = evalPath(segments, piece, cn);
  if (rings.length < 2) return null;

  const vertices: V3[] = [];
  for (const ring of rings) vertices.push(...ring.v);

  const faces: number[][] = [];
  for (let i = 0; i < rings.length - 1; i += 1) {
    const a0 = i * cn;
    const b0 = (i + 1) * cn;
    for (let j = 0; j < cn; j += 1) {
      const jn = (j + 1) % cn;
      faces.push([a0 + j, a0 + jn, b0 + jn]);
      faces.push([a0 + j, b0 + jn, b0 + j]);
    }
  }

  for (const [ring, rbase, start] of [[rings[0], 0, true], [rings[rings.length - 1], (rings.length - 1) * cn, false]] as const) {
    const cap = capMesh(ring, piece, cn, start, resolution);
    if (!cap) continue;
    const offset = vertices.length;
    vertices.push(...cap.vertices as V3[]);
    for (const face of cap.faces) faces.push([face[0] + offset, face[1] + offset, face[2] + offset]);
    connectCap(faces, cap.vertices, offset, rbase, cn, start);
  }

  return faces.length ? { vertices, faces } : null;
}

function seg(points: Point[], i: number, piece: ResolvedPiece, resolution: number): Segment | null {
  const p0: V3 = [points[i].x, points[i].y, points[i].z];
  const p3: V3 = [points[i + 1].x, points[i + 1].y, points[i + 1].z];
  const co = points[i].curve_out;
  const ci = points[i + 1].curve_in;
  const p1: V3 = co ? [co.x, co.y, co.z] : [...p0];
  const p2: V3 = ci ? [ci.x, ci.y, ci.z] : [...p3];
  const linear = near(p1, p0) && near(p2, p3);

  if (![0, 0.25, 0.5, 0.75, 1].some((t) => norm(cubicBezierDeriv(p0, p1, p2, p3, t)) > 1e-12)) return null;

  const s0 = effectiveSize(points[i], piece);
  const s1 = effectiveSize(points[i + 1], piece);
  const r0 = effectiveRotation(points[i], piece);
  const r1 = effectiveRotation(points[i + 1], piece);
  const to = points[i].transition_out;
  const ti = points[i + 1].transition_in;
  const t1: [number, number] = to ? [to.x, to.y] : [1 / 3, 1 / 3];
  const t2: [number, number] = ti ? [ti.x, ti.y] : [2 / 3, 2 / 3];
  const hasTransition = Boolean(to || ti);
  const steps = linear ? Math.max(1, Math.floor(resolution / 4)) : resolution;

  const position = (u: number) => linear ? add(scale(p0, 1 - u), scale(p3, u)) : cubicBezier(p0, p1, p2, p3, u);
  const linDir = sub(p3, p0);
  const tangent = (u: number) => {
    if (linear) return linDir;
    const deriv = cubicBezierDeriv(p0, p1, p2, p3, u);
    if (norm(deriv) < 1e-12) {
      for (const delta of [0.01, -0.01, 0.05, -0.05, 0.1, -0.1]) {
        const t = Math.max(0, Math.min(1, u + delta));
        const d2 = cubicBezierDeriv(p0, p1, p2, p3, t);
        if (norm(d2) > 1e-12) return d2;
      }
    }
    return deriv;
  };
  const v = (u: number) => hasTransition ? solveTransition(t1, t2, u) : u;
  const size = (u: number): [number, number, number] => interpolateSize({ x: s0[0], y: s0[1], z: s0[2] }, { x: s1[0], y: s1[1], z: s1[2] }, v(u));
  const rot = (u: number): [number, number, number] => interpolateRotation({ x: r0[0], y: r0[1], z: r0[2] }, { x: r1[0], y: r1[1], z: r1[2] }, v(u));
  return { n: steps, pos: position, tan: tangent, size, rot };
}

function evalPath(segments: Segment[], piece: ResolvedPiece, cn: number): Ring[] {
  const rings: Ring[] = [];
  let previousFrame: Frame | null = null;
  let previousTangent: V3 | null = null;

  for (let si = 0; si < segments.length; si += 1) {
    const segment = segments[si];
    for (let step = 0; step <= segment.n; step += 1) {
      if (si > 0 && step === 0) continue;
      const u = step / segment.n;
      const pos = segment.pos(u);
      const rawTangent = segment.tan(u);
      if (norm(rawTangent) < 1e-12) continue;
      const tangent = normalize(rawTangent);
      const size = segment.size(u);
      const isEndpoint = (si === 0 && step === 0) || (si === segments.length - 1 && step === segment.n);
      if (!isEndpoint && degenerate(...size)) continue;

      const rotation = rotationMatrixXYZ(...segment.rot(u));
      const frame: Frame = previousFrame === null ? initFrame(tangent, rotation) : transport(previousFrame, previousTangent!, tangent, rotation);
      previousFrame = frame;
      previousTangent = tangent;

      rings.push({ pos, frame, size, rot: segment.rot(u), v: crossSection(pos, frame, size, piece.shape, cn) });
    }
  }
  return rings;
}

function initFrame(tangent: V3, rotation: Mat3): Frame {
  const lz = tangent;
  const rx = matVec(rotation, [1, 0, 0]);
  const px = projectOntoPlane(rx, lz);
  const lx = norm(px) > 1e-10 ? normalize(px) : normalize(projectOntoPlane(matVec(rotation, [0, 1, 0]), lz));
  const ly = normalize(cross(lz, lx));
  return { lx, ly, lz };
}

function transport(prev: Frame, previousTangent: V3, currentTangent: V3, rotation: Mat3): Frame {
  const d = dot(previousTangent, currentTangent);
  if (d > 1 - 1e-10) return { lx: [...prev.lx], ly: [...prev.ly], lz: currentTangent };
  if (d < -1 + 1e-10) return initFrame(currentTangent, rotation);
  const r = minimalRotation(previousTangent, currentTangent);
  return { lx: normalize(matVec(r, prev.lx)), ly: normalize(matVec(r, prev.ly)), lz: currentTangent };
}

function crossSection(pos: V3, frame: Frame, size: [number, number, number], shape: string, n: number): V3[] {
  void shape;
  const [sx, sy] = size;
  return Array.from({ length: n }, (_, i) => {
    const angle = 2 * Math.PI * i / n;
    return add(add(pos, scale(frame.lx, sx * Math.cos(angle))), scale(frame.ly, sy * Math.sin(angle)));
  });
}

function capMesh(ring: Ring, piece: ResolvedPiece, cn: number, isStart: boolean, resolution: number): MeshData | null {
  const sz = ring.size[2];
  if (sz === 0) return { vertices: [ring.pos], faces: [] };
  return roundedCap(ring, piece, cn, isStart, resolution);
}

function roundedCap(ring: Ring, piece: ResolvedPiece, cn: number, isStart: boolean, resolution: number): MeshData {
  const { pos, frame } = ring;
  const [sx, sy, sz] = ring.size;
  const d = isStart ? -1 : 1;
  if (piece.shape === 'circle') {
    const nl = Math.max(2, Math.floor(resolution / 4));
    const vertices: V3[] = [add(pos, scale(frame.lz, d * sz))];
    for (let i = 1; i < nl; i += 1) {
      const phi = Math.PI / 2 * i / nl;
      const sp = Math.sin(phi);
      const cp = Math.cos(phi);
      for (let j = 0; j < cn; j += 1) {
        const angle = 2 * Math.PI * j / cn;
        vertices.push(add(add(add(pos, scale(frame.lx, sx * sp * Math.cos(angle))), scale(frame.ly, sy * sp * Math.sin(angle))), scale(frame.lz, sz * cp * d)));
      }
    }
    const faces: number[][] = [];
    for (let j = 0; j < cn; j += 1) {
      const jn = (j + 1) % cn;
      faces.push(isStart ? [0, 1 + jn, 1 + j] : [0, 1 + j, 1 + jn]);
    }
    for (let i = 0; i < nl - 2; i += 1) {
      const a0 = 1 + i * cn;
      const b0 = 1 + (i + 1) * cn;
      for (let j = 0; j < cn; j += 1) {
        const jn = (j + 1) % cn;
        if (isStart) {
          faces.push([a0 + j, a0 + jn, b0 + jn]);
          faces.push([a0 + j, b0 + jn, b0 + j]);
        } else {
          faces.push([a0 + j, b0 + j, b0 + jn]);
          faces.push([a0 + j, b0 + jn, a0 + jn]);
        }
      }
    }
    return { vertices, faces };
  }
  return { vertices: [add(pos, scale(frame.lz, d * sz))], faces: [] };
}

function connectCap(faces: number[][], capVertices: number[][], offset: number, rbase: number, cn: number, isStart: boolean): void {
  const nv = capVertices.length;
  if (nv === 1) {
    for (let j = 0; j < cn; j += 1) {
      const jn = (j + 1) % cn;
      const a = rbase + j;
      const b = rbase + jn;
      faces.push(isStart ? [offset, b, a] : [offset, a, b]);
    }
    return;
  }
  const nr = Math.floor((nv - 1) / cn);
  if (nr > 0) {
    const lr = offset + 1 + (nr - 1) * cn;
    for (let j = 0; j < cn; j += 1) {
      const jn = (j + 1) % cn;
      const a = lr + j;
      const b = lr + jn;
      const c = rbase + jn;
      const d = rbase + j;
      if (isStart) {
        faces.push([a, b, c]);
        faces.push([a, c, d]);
      } else {
        faces.push([a, d, c]);
        faces.push([a, c, b]);
      }
    }
  } else {
    for (let j = 0; j < cn; j += 1) {
      const jn = (j + 1) % cn;
      const a = rbase + j;
      const b = rbase + jn;
      faces.push(isStart ? [offset, b, a] : [offset, a, b]);
    }
  }
}

function near(a: V3, b: V3): boolean {
  return Math.abs(a[0] - b[0]) <= 1e-8 && Math.abs(a[1] - b[1]) <= 1e-8 && Math.abs(a[2] - b[2]) <= 1e-8;
}

File: implementation/packages/python/ssklib/tessellate.py
import math

import numpy as np

from .error import SSKError
from .vecmath import (
    vec3, normalize, rotation_matrix_xyz, cubic_bezier, cubic_bezier_deriv,
    solve_transition, minimal_rotation, interpolate_rotation,
    interpolate_size, project_onto_plane,
)


_DEFAULT_RESOLUTION = 32
_MIN_RESOLUTION = 3


def tessellate(piece: dict, *, resolution: int = _DEFAULT_RESOLUTION):
    resolution = _check_resolution(resolution)

    if len(piece['points']) == 1:
        return _point_defined(piece, resolution)
    return _path_defined(piece, resolution)


def _check_resolution(resolution: int) -> int:
    if isinstance(resolution, bool) or not isinstance(resolution, int):
        raise SSKError("resolution must be an integer")
    if resolution < _MIN_RESOLUTION:
        raise SSKError(f"resolution must be >= {_MIN_RESOLUTION}")
    return resolution


def _eff_size(pt: dict, piece: dict) -> tuple:
    s = pt.get('size', piece['size'])
    return (float(s['x']), float(s['y']), float(s['z']))


def _eff_rot(pt: dict, piece: dict) -> tuple:
    r = pt.get('rotation', piece.get('rotation', {'x': 0.0, 'y': 0.0, 'z': 0.0}))
    return (float(r['x']), float(r['y']), float(r['z']))


def _degenerate(sx, sy, sz) -> bool:
    return sum(1 for v in (sx, sy, sz) if v == 0.0) >= 2


def _cross_n(piece: dict, resolution: int) -> int:
    return piece.get('sides', resolution) if piece['shape'] == 'ngon' else resolution


# point-defined piece  (spec/geometry/SPEC.md 8.2)

def _point_defined(piece: dict, resolution: int):
    pt = piece['points'][0]
    sx, sy, sz = _eff_size(pt, piece)
    if _degenerate(sx, sy, sz):
        return None, None
    pos = vec3(pt['x'], pt['y'], pt['z'])
    R = rotation_matrix_xyz(*_eff_rot(pt, piece))
    if piece['shape'] == 'circle':
        return _ellipsoid(pos, sx, sy, sz, R, resolution)
    return _bipyramid(pos, sx, sy, sz, R, piece['sides'])


def _ellipsoid(c, sx, sy, sz, R, n):
    n_lat, n_lon = max(4, n // 2), max(6, n)
    vs = [c + R @ vec3(0, 0, sz)]
    for i in range(1, n_lat):
        phi = math.pi * i / n_lat
        sp, cp = math.sin(phi), math.cos(phi)
        for j in range(n_lon):
            th = 2 * math.pi * j / n_lon
            vs.append(c + R @ vec3(sx * sp * math.cos(th),
                                   sy * sp * math.sin(th), sz * cp))
    vs.append(c + R @ vec3(0, 0, -sz))

    fs = []
    for j in range(n_lon):
        fs.append([0, 1 + j, 1 + (j + 1) % n_lon])
    for i in range(n_lat - 2):
        a0, b0 = 1 + i * n_lon, 1 + (i + 1) * n_lon
        for j in range(n_lon):
            jn = (j + 1) % n_lon
            fs.append([a0 + j, b0 + j, a0 + jn])
            fs.append([a0 + jn, b0 + j, b0 + jn])
    bot = 1 + (n_lat - 1) * n_lon
    last = 1 + (n_lat - 2) * n_lon
    for j in range(n_lon):
        fs.append([last + j, bot, last + (j + 1) % n_lon])
    return np.array(vs, dtype=np.float64), np.array(fs, dtype=np.int32)


def _bipyramid(c, sx, sy, sz, R, sides):
    vs = [c + R @ vec3(0, 0, sz), c + R @ vec3(0, 0, -sz)]
    for i in range(sides):
        a = 2 * math.pi * i / sides
        vs.append(c + R @ vec3(sx * math.cos(a), sy * math.sin(a), 0))
    fs = []
    for i in range(sides):
        n = (i + 1) % sides
        fs.append([0, 2 + i, 2 + n])
        fs.append([1, 2 + n, 2 + i])
    return np.array(vs, dtype=np.float64), np.array(fs, dtype=np.int32)


# path-defined piece  (spec/geometry/SPEC.md 8.1)

def _path_defined(piece: dict, resolution: int):
    points = piece['points']
    cn = _cross_n(piece, resolution)

    segs = [s for i in range(len(points) - 1)
            if (s := _seg(points, i, piece, resolution)) is not None]
    if not segs:
        return None, None

    rings = _eval_path(segs, piece, cn)
    if len(rings) < 2:
        return None, None

    body_v = []
    for r in rings:
        body_v.extend(r['v'])
    verts = np.array(body_v, dtype=np.float64)

    faces = []
    nr = len(rings)
    for i in range(nr - 1):
        a0, b0 = i * cn, (i + 1) * cn
        for j in range(cn):
            jn = (j + 1) % cn
            faces.append([a0 + j, a0 + jn, b0 + jn])
            faces.append([a0 + j, b0 + jn, b0 + j])

    for ring, rbase, start in [(rings[0], 0, True),
                                (rings[-1], (nr - 1) * cn, False)]:
        cap = _cap(ring, piece, cn, start, resolution)
        if cap is None:
            continue
        cv, cf = cap
        off = len(verts)
        if len(cv):
            verts = np.vstack([verts, cv])
        for f in cf:
            faces.append([f[0] + off, f[1] + off, f[2] + off])
        _connect_cap(faces, cv, off, rbase, cn, start)

    if not faces:
        return None, None
    return verts, np.array(faces, dtype=np.int32)


def _seg(points, i, piece, resolution: int):

    p0 = vec3(points[i]['x'], points[i]['y'], points[i]['z'])
    p3 = vec3(points[i + 1]['x'], points[i + 1]['y'], points[i + 1]['z'])

    co = points[i].get('curve_out')
    p1 = vec3(co['x'], co['y'], co['z']) if co else p0.copy()
    ci = points[i + 1].get('curve_in')
    p2 = vec3(ci['x'], ci['y'], ci['z']) if ci else p3.copy()

    linear = np.allclose(p1, p0) and np.allclose(p2, p3)

    # reject degenerate segments (7.1)
    if not any(np.linalg.norm(cubic_bezier_deriv(p0, p1, p2, p3, t)) > 1e-12
               for t in (0, .25, .5, .75, 1)):
        return None

    s0, s1 = _eff_size(points[i], piece), _eff_size(points[i + 1], piece)
    r0, r1 = _eff_rot(points[i], piece),  _eff_rot(points[i + 1], piece)

    to = points[i].get('transition_out')
    ti = points[i + 1].get('transition_in')
    t1 = (to['x'], to['y']) if to else (1/3, 1/3)
    t2 = (ti['x'], ti['y']) if ti else (2/3, 2/3)
    has_t = to is not None or ti is not None

    steps = resolution if not linear else max(1, resolution // 4)

    def pos(u):
        if linear:
            return (1.0 - u) * p0 + u * p3
        return cubic_bezier(p0, p1, p2, p3, u)

    _lin_dir = p3 - p0

    def tan(u):
        if linear:
            return _lin_dir
        d = cubic_bezier_deriv(p0, p1, p2, p3, u)
        if np.linalg.norm(d) < 1e-12:
            for dt in (.01, -.01, .05, -.05, .1, -.1):
                d2 = cubic_bezier_deriv(p0, p1, p2, p3, max(0, min(1, u + dt)))
                if np.linalg.norm(d2) > 1e-12:
                    return d2
        return d

    def _v(u):
        return solve_transition(t1, t2, u) if has_t else u

    def size(u):
        v = _v(u)
        return interpolate_size({'x': s0[0], 'y': s0[1], 'z': s0[2]},
                                {'x': s1[0], 'y': s1[1], 'z': s1[2]}, v)

    def rot(u):
        v = _v(u)
        return interpolate_rotation({'x': r0[0], 'y': r0[1], 'z': r0[2]},
                                    {'x': r1[0], 'y': r1[1], 'z': r1[2]}, v)

    return {'n': steps, 'pos': pos, 'tan': tan, 'size': size, 'rot': rot}


def _eval_path(segs, piece, cn):

    rings = []
    pf = pt = None  # previous frame, previous tangent

    for si, seg in enumerate(segs):
        for step in range(seg['n'] + 1):
            if si > 0 and step == 0:
                continue
            u = step / seg['n']
            p = seg['pos'](u)
            t_raw = seg['tan'](u)
            if np.linalg.norm(t_raw) < 1e-12:
                continue
            t = normalize(t_raw)
            sz = seg['size'](u)
            is_endpoint = (si == 0 and step == 0) or (si == len(segs) - 1 and step == seg['n'])
            if not is_endpoint and _degenerate(*sz):
                continue

            Rr = rotation_matrix_xyz(*seg['rot'](u))
            if pf is None:
                frame = _init_frame(t, Rr)
            else:
                frame = _transport(pf, pt, t, Rr)
            pf, pt = frame, t

            rv = _cross(p, frame, sz, piece['shape'], cn)
            rings.append({'pos': p, 'frame': frame, 'size': sz,
                          'rot': seg['rot'](u), 'v': rv})
    return rings


# frame construction  (spec/geometry/SPEC.md 9.1)

def _init_frame(tangent, R_rot):
    lz = tangent
    rx = R_rot @ vec3(1, 0, 0)
    px = project_onto_plane(rx, lz)
    if np.linalg.norm(px) > 1e-10:
        lx = normalize(px)
    else:
        lx = normalize(project_onto_plane(R_rot @ vec3(0, 1, 0), lz))
    ly = normalize(np.cross(lz, lx))
    return np.column_stack([lx, ly, lz])


def _transport(prev, pt, ct, R_rot):
    d = float(np.dot(pt, ct))
    if d > 1.0 - 1e-10:
        f = prev.copy(); f[:, 2] = ct; return f
    if d < -1.0 + 1e-10:
        return _init_frame(ct, R_rot)
    R = minimal_rotation(pt, ct)
    return np.column_stack([normalize(R @ prev[:, 0]),
                            normalize(R @ prev[:, 1]), ct])


def _cross(pos, frame, size, shape, n):
    sx, sy = size[0], size[1]
    lx, ly = frame[:, 0], frame[:, 1]
    return [pos + sx * math.cos(2 * math.pi * i / n) * lx
                + sy * math.sin(2 * math.pi * i / n) * ly
            for i in range(n)]


# caps  (spec/geometry/SPEC.md 8.3)

def _cap(ring, piece, cn, is_start, resolution: int):
    sz = ring['size'][2]
    if sz == 0.0:
        return np.array([ring['pos']], dtype=np.float64), []
    return _rounded_cap(ring, piece, cn, is_start, resolution)


def _rounded_cap(ring, piece, cn, is_start, resolution: int):
    pos = ring['pos']
    lx, ly, lz = ring['frame'][:, 0], ring['frame'][:, 1], ring['frame'][:, 2]
    sx, sy, sz = ring['size']
    d = -1.0 if is_start else 1.0

    if piece['shape'] == 'circle':
        nl = max(2, resolution // 4)
        vs = [pos + d * sz * lz]
        for i in range(1, nl):
            phi = math.pi / 2 * i / nl
            sp, cp = math.sin(phi), math.cos(phi)
            for j in range(cn):
                a = 2 * math.pi * j / cn
                vs.append(pos + sx * sp * math.cos(a) * lx
                              + sy * sp * math.sin(a) * ly
                              + sz * cp * d * lz)
        fs = []
        for j in range(cn):
            jn = (j + 1) % cn
            fs.append([0, 1 + jn, 1 + j] if is_start else [0, 1 + j, 1 + jn])
        for i in range(nl - 2):
            a0, b0 = 1 + i * cn, 1 + (i + 1) * cn
            for j in range(cn):
                jn = (j + 1) % cn
                if is_start:
                    fs.append([a0 + j, a0 + jn, b0 + jn])
                    fs.append([a0 + j, b0 + jn, b0 + j])
                else:
                    fs.append([a0 + j, b0 + j, b0 + jn])
                    fs.append([a0 + j, b0 + jn, a0 + jn])
        return np.array(vs, dtype=np.float64), fs

    # ngon half-bipyramid: apex only
    return np.array([pos + d * sz * lz], dtype=np.float64), []


def _connect_cap(faces, cv, off, rbase, cn, is_start):
    nv = len(cv)
    if nv == 1:
        apex = off
        for j in range(cn):
            jn = (j + 1) % cn
            a, b = rbase + j, rbase + jn
            faces.append([apex, b, a] if is_start else [apex, a, b])
        return
    nr = (nv - 1) // cn
    if nr > 0:
        lr = off + 1 + (nr - 1) * cn
        for j in range(cn):
            jn = (j + 1) % cn
            a, b, c, d = lr + j, lr + jn, rbase + jn, rbase + j
            if is_start:
                faces.append([a, b, c]); faces.append([a, c, d])
            else:
                faces.append([a, d, c]); faces.append([a, c, b])
    else:
        apex = off
        for j in range(cn):
            jn = (j + 1) % cn
            a, b = rbase + j, rbase + jn
            faces.append([apex, b, a] if is_start else [apex, a, b])
